[Caffe2] Add SVE128 vectorized<float> template layer

Nicoshev opened the pull request:
Summary:

This PR is part of a set of changes that aim to introduce SVE128 as a separate VecLib ISA.

Motivation is that VecLib's aarch64 support is much more limited than x86
Context is that PyTorch's current SVE ISA is not meant to be run on CPUs with vector length lower than 256-bits, and NEON ISA does not support doubles, ints, uints. While x86 does support them, and more.

Goal is to leave platform parity between modern aarch64 and x86 systems.

In this PR, we are just adding the float translation layer. That is, support for Vectorized<float>

One decision within these changes is whether to reuse or not existing NEON code for the currently supported types, like float.

We opted for writing new float/f16/bf16 translation layers because:
  1. It provides a small performance improvement in some scenarios
  2. Having a completely separate modern platform allows us to tweak other things, like compilation settings, without affecting older CPUs using NEON

Passes test_ops.py

Test Plan:
Sigrid Predictor

Rollback Plan:

Differential Revision: D79199881


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @malfet @snadampal @milpuz01 @aditew01 @nikhil-arm @fadara01 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

Skylion007 commented on aten/src/ATen/cpu/vec/sve/vec_double.h:
```diff
@@ -178,6 +173,180 @@ class Vectorized<double> {
+  Vectorized<double> atan2(const Vectorized<double>& b) const {USE_SLEEF(
+      { return Vectorized<double>(Sleef_atan2d2_u10(svget_neonq(values), b)); },
+      {
+        __at_align__ double tmp[size()];
```
Nit: std::array?

malfet reviewed (commented):
Can you please elaborate why vec128_float_sve can not share some/any code with vec256_float_sve?

malfet commented:
@aditew01 do you mind having a look at it?

malfet commented on aten/src/ATen/cpu/vec/vec128/vec128_float_sve.h:
```diff
@@ -0,0 +1,778 @@
+#include <c10/util/irange.h>
+#include <cmath>
+
+#ifdef USE_FBGEMM
+C10_DIAGNOSTIC_PUSH_AND_IGNORED_IF_DEFINED("-Wextra-semi")
+#include <fbgemm/Fbgemm.h>
+C10_DIAGNOSTIC_POP()
+#endif
```
Sorry what is that? Why this depends on FBGEMM?

Nicoshev replied:
@malfet To use FBGEMM's transpose_mxn function. We've worked with ARM the company to land an assembly-based routine that is very fast. Thought that reusing such code was better than copying it here.

malfet reviewed (changes requested):
VecLib should be a header only library, that might depend on sleef, but defintely not on fbgemm

malfet reviewed (changes requested):
Regarding test plan: 
Considering that this is performance rather than functional change, running functional test correctness is not sufficient. Please provide a benchmark

`vec128_transpose.h` seems to be adding NEON rather that SVE functionality and with that in mind, I'm not sure what it has to do with PR title

aditew01 commented on aten/src/ATen/cpu/vec/vec128/vec128_float_sve.h:
```diff
@@ -0,0 +1,769 @@
+template <>
+class Vectorized<float> {
+ private:
+  float32x4_t values;
```
Is this to skip the compile issues ?

Nicoshev committed:
[Caffe2] Add SVE128 vectorized<float> template layer (#160328)

Summary:

We are introducing the SVE128 vectorized<> layer.

Idea is to mix NEON and SVE to maximize performance on SVE128 cpus

This first diff adds vectorized<float>

Test Plan: Sigrid Predictor

Reviewed By: mcfi

Differential Revision: D79199881
